pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,494 @@
+					if err != nil {
+						level.Error(util.Logger).Log("msg", "error removing delete plan", "request-id", jobExecutionStatus.deleteRequestID, "plan-no", jobExecutionStatus.planNo, "err", err)
+					} else {
+						pendingPlansCountMtx.Lock()
+						pendingPlansCount[jobExecutionStatus.deleteRequestID]--
+
+						if pendingPlansCount[jobExecutionStatus.deleteRequestID] == 0 {
+							level.Debug(util.Logger).Log("msg", "finished execution of all plans so cleaning up and updating status of request",
+								"request-id", jobExecutionStatus.deleteRequestID, "user-id", jobExecutionStatus.userID)
+
+							err := dp.deleteStore.UpdateStatus(context.Background(), jobExecutionStatus.userID, jobExecutionStatus.deleteRequestID, chunk.DeleteRequestStatusProcessed)
+							if err != nil {
+								level.Error(util.Logger).Log("msg", "error updating delete request status to process", "request-id", jobExecutionStatus.deleteRequestID, "err", err)
+							}
+
+							delete(pendingPlansCount, jobExecutionStatus.deleteRequestID)
+							pendingPlansCountMtx.Unlock()
+
+							dp.inProcessRequestIDsMtx.Lock()
+							delete(dp.inProcessRequestIDs, jobExecutionStatus.deleteRequestID)
+							dp.inProcessRequestIDsMtx.Unlock()
+						} else {
+							pendingPlansCountMtx.Unlock()
+
+						}
+
+					}
+				}
+			case <-dp.quit:
+				dp.wg.Done()
+				break
+			}
+		}
+	}()
+
+	dp.wg.Add(1)
+
+	for {
+		select {
+		case deleteRequest := <-dp.executePlansChan:
+			numPlans := numOfPlans(deleteRequest.StartTime, deleteRequest.EndTime)
+			level.Debug(util.Logger).Log("msg", "sending jobs to workers for purging data",
+				"request-id", deleteRequest.RequestID, "user-id", deleteRequest.UserID, "num-jobs", numPlans)
+
+			pendingPlansCountMtx.Lock()
+			pendingPlansCount[deleteRequest.RequestID] = numPlans
+			pendingPlansCountMtx.Unlock()
+
+			for i := 0; i < numPlans; i++ {
+				dp.workerJobChan <- workerJob{planNo: i, userID: deleteRequest.UserID, deleteRequestID: deleteRequest.RequestID}
+			}
+		case <-dp.quit:
+			close(dp.workerJobChan)
+			dp.wg.Done()
+			break
+		}
+	}
+}
+
+func (dp *DataPurger) runWorkers() {
+	for i := 0; i < dp.cfg.NumWorkers; i++ {
+		go dp.worker()
+	}
+}
+
+func (dp *DataPurger) worker() {
+	for job := range dp.workerJobChan {
+		err := dp.executePlan(job.userID, job.deleteRequestID, job.planNo)
+		dp.workerJobExecutionStatusChan <- workerJobExecutionStatus{job, err}
+	}
+}
+
+func (dp *DataPurger) executePlan(userID, requestID string, planNo int) error {
+	pb, err := dp.deleteStore.GetDeletePlan(context.Background(), userID, requestID, planNo)
+	if err != nil {
+		return err
+	}
+
+	if pb == nil {
+		level.Debug(util.Logger).Log("msg", "plan not found, must be executed already",
+			"request-id", requestID, "user-id", userID, "plan-no", planNo)
+		// this means plan was already executed and got removed. Do nothing.
+		return nil
+	}
+
+	level.Debug(util.Logger).Log("msg", "executing plan",
```

Comment:
I didn't want too much noise in logs so I used info level only for important logs since a delete request could touch thousands of series and chunks. Will go through them again and switch some of them to info level until it gets stable.

Reply:
I think any time a plan is being exectuted/completed we should log.